### ISSUE TITLE
Proposal: Dockerfile restructuring

### ISSUE THREAD

**Issue by tgraf** [cilium/cilium#1998]
**Requirements**
* Build
   * Build cilium binaries based on golang code
   * Build `cilium-envoy` binary based on C++ code
* Runtime image
  * Minimal in size
  * Needs to contain clang + iproute2
  * Does not require golang + C++ + bazel build environment
  * Minimal debugging tools

**Proposal**
 - [x] `cilium/build-base` image: golang + bazel + protobuf + precompiled envoy
 - [x] `cilium/runtime-base`: ubuntu + clang + iproute2 + gops
 - [x] Dockerfile: Multistage build
    - [x] Stage 1: `FROM cilium/build-base`, build
    - [ ] Stage 2: `FROM cilium/runtime-base`, copies binaries from stage1

**Comment by ianvernon** [comment]
This looks great! We will need a newer version of Docker to do multistage builds than what is currently in the Vagrant VM. I am addressing this as part of the Ginkgo CI migration, in which I install Docker CE 17.06

**Comment by nebril** [comment]
Can we evaluate using Alpine for runtime image?

clang package: https://pkgs.alpinelinux.org/package/edge/main/x86/clang
iproute2: https://pkgs.alpinelinux.org/package/edge/main/x86/iproute2

I wonder if we need everything from clang package, not sure TBH.

**Comment by tgraf** [comment]
> Can we evaluate using Alpine for runtime image?
> 
> clang package: https://pkgs.alpinelinux.org/package/edge/main/x86/clang
> iproute2: https://pkgs.alpinelinux.org/package/edge/main/x86/iproute2
> 
> I wonder if we need everything from clang package, not sure TBH.

Given the new multi stage builds, exchanging the base image will become very simple. We already slim down the clang package to a minimum.

**Comment by tgraf** [comment]
I've completed the build base image which includes a fully built envoy in `/root/.cache/`.
Unfortunately it's huge in size:
```
cilium/build        latest              004b8c554398        6 minutes ago       4.45GB
```

**Comment by aanm** [comment]
@tgraf did you try to compress the objects in `/roo/.cache` in a tarball? Then we would get that tarball from in the multi stage for the runtime image.

**Comment by tgraf** [comment]
> did you try to compress the objects in /roo/.cache in a tarball? Then we would get that tarball from in the multi stage for the runtime image.

The tarball will reduce the image size but it will also dramatically increase the build time because it has to be unpacked/uncompressed every time.

**Comment by ianvernon** [comment]
@jrajahalme @tgraf can this be closed since we can do multi-stage builds with Docker now?

**Comment by tgraf** [comment]
We haven't done a new multi stage based runtime yet. I have this locally and can merge it with Jarno's multi stage changes.

**Comment by tgraf** [comment]
Implemented since 1.3
